src/lua/utils.c

```diff
@@ -735,6 +735,22 @@ luaT_newthread(struct lua_State *L)
 	return L1;
 }
 
+/** A getenv wrapper checking for size of the corresponding env variable. */
+static int
+luaT_getenv_safe(struct lua_State *L)
+{
+	if (lua_gettop(L) != 2 || !lua_isstring(L, 1) || !lua_isnumber(L, 2))
+		return luaL_error(L, "usage: getenv_safe(name, length)");
+	const char *name = lua_tostring(L, 1);
+	int len = lua_tonumber(L, 2);
+	const char *env = getenv_safe(name, len, NULL);
+	if (env != NULL) {
+		lua_pushstring(L, env);
+		return 1;
+	}
+	return 0;
+}
```

Comment:
We have `src/lua/env.lua`. Maybe it'll be more appropriate to add `src/lua/env.c`?

Just raw idea.

Reply:
Even better, I moved the function to `env.lua`:
<details><summary>diff</summary>

```diff
diff --git a/extra/exports b/extra/exports
index f30e7bbfe..a068bafb8 100644
--- a/extra/exports
+++ b/extra/exports
@@ -227,6 +227,7 @@ fiber_time64
 fiber_wakeup
 fiber_yield
 guava
+getenv_safe
 ibuf_create
 ibuf_destroy
 ibuf_reinit
diff --git a/src/lua/env.lua b/src/lua/env.lua
index a31b7098f..9a505c6b3 100644
--- a/src/lua/env.lua
+++ b/src/lua/env.lua
@@ -7,6 +7,7 @@ ffi.cdef[[
 
     int   setenv(const char *name, const char *value, int overwrite);
     int   unsetenv(const char *name);
+    char * getenv_safe(const char *name, size_t buf_size, bool *exceeded);
 ]]
 
 os.environ = function()
@@ -38,3 +39,12 @@ os.setenv = function(key, value)
         error(string.format('error %d: %s', errno(), errno.errstring()))
     end
 end
+
+os.getenv_safe = function(key, size)
+    local ret = ffi.C.getenv_safe(key, size, nil)
+    if ret ~= nil then
+        return ffi.string(ret)
+    else
+        return nil
+    end
+end
diff --git a/src/lua/utils.c b/src/lua/utils.c
index 9b3ac9400..49858dd1b 100644
--- a/src/lua/utils.c
+++ b/src/lua/utils.c
@@ -735,22 +735,6 @@ luaT_newthread(struct lua_State *L)
 	return L1;
 }
 
-/** A getenv wrapper checking for size of the corresponding env variable. */
-static int
-luaT_getenv_safe(struct lua_State *L)
-{
-	if (lua_gettop(L) != 2 || !lua_isstring(L, 1) || !lua_isnumber(L, 2))
-		return luaL_error(L, "usage: getenv_safe(name, length)");
-	const char *name = lua_tostring(L, 1);
-	int len = lua_tonumber(L, 2);
-	const char *env = getenv_safe(name, len, NULL);
-	if (env != NULL) {
-		lua_pushstring(L, env);
-		return 1;
-	}
-	return 0;
-}
-
 int
 tarantool_lua_utils_init(struct lua_State *L)
 {
@@ -816,10 +800,8 @@ tarantool_lua_utils_init(struct lua_State *L)
 	CTID_INTERVAL = luaL_ctypeid(L, "struct interval");
 	assert(CTID_INTERVAL != 0);
 
-	/* os.getenv_safe() */
+	/* os.PATH_MAX */
 	lua_getglobal(L, "os");
-	lua_pushcfunction(L, luaT_getenv_safe);
-	lua_setfield(L, -2, "getenv_safe");
 	lua_pushnumber(L, PATH_MAX);
 	lua_setfield(L, -2, "PATH_MAX");
 	lua_pop(L, 1);
diff --git a/src/trivia/util.h b/src/trivia/util.h
index 311c898e3..2309a9e75 100644
--- a/src/trivia/util.h
+++ b/src/trivia/util.h
@@ -648,7 +648,7 @@ thread_sleep(double sec);
  * buffer of the given size. That is, its length should be <= buffer size - 1,
  * because one extra char is taken by the trailing '\0'.
  */
-char *
+API_EXPORT char *
 getenv_safe(const char *name, size_t buf_size, bool *exceeded);
 
 #if !defined(__cplusplus) && !defined(static_assert)
```
</details>

Reply:
I'll not insist on it here, but I had an idea to eliminate exports, which are for internal usage using `tnt_internal_symbol()` or a similar mechanism. Those are not actually public functions, so it is undesirable to export them.

Reply:
Looks scary, but cool:
```diff
diff --git a/extra/exports b/extra/exports
index a068bafb8..f30e7bbfe 100644
--- a/extra/exports
+++ b/extra/exports
@@ -227,7 +227,6 @@ fiber_time64
 fiber_wakeup
 fiber_yield
 guava
-getenv_safe
 ibuf_create
 ibuf_destroy
 ibuf_reinit
diff --git a/src/lua/env.lua b/src/lua/env.lua
index 059d6b5d7..d6f92bb7d 100644
--- a/src/lua/env.lua
+++ b/src/lua/env.lua
@@ -7,7 +7,7 @@ ffi.cdef[[
 
     int   setenv(const char *name, const char *value, int overwrite);
     int   unsetenv(const char *name);
-    char * getenv_safe(const char *name, size_t buf_size, bool *exceeded);
+    void *tnt_internal_symbol(const char *name);
 ]]
 
 os.environ = function()
@@ -40,6 +40,9 @@ os.setenv = function(key, value)
     end
 end
 
+local getenv_safe = ffi.cast('char * (*)(const char *, size_t, bool *)',
+                             ffi.C.tnt_internal_symbol('getenv_safe'))
+
 --
 -- Return the environment variable pointed to by key, as long as its length is
 -- strictly less than size.
@@ -49,7 +52,7 @@ end
 -- to pass the value to C).
 --
 os.getenv_safe = function(key, size)
-    local ret = ffi.C.getenv_safe(key, size, nil)
+    local ret = getenv_safe(key, size, nil)
     if ret ~= nil then
         return ffi.string(ret)
     else
diff --git a/src/symbols.c b/src/symbols.c
index 98bba729c..92b6b1bb8 100644
--- a/src/symbols.c
+++ b/src/symbols.c
@@ -45,6 +45,8 @@ extern void
 ipc_value_delete(void);
 extern void
 ipc_value_new(void);
+extern void
+getenv_safe(void);
 
 /** Symbol definition. */
 struct symbol_def {
@@ -68,6 +70,7 @@ static struct symbol_def symbols[] = {
 	{"fiber_channel_put_timeout", fiber_channel_put_timeout},
 	{"ipc_value_delete", ipc_value_delete},
 	{"ipc_value_new", ipc_value_new},
+	{"getenv_safe", getenv_safe},
 	{NULL, NULL}
 };
 
diff --git a/src/trivia/util.h b/src/trivia/util.h
index 2309a9e75..311c898e3 100644
--- a/src/trivia/util.h
+++ b/src/trivia/util.h
@@ -648,7 +648,7 @@ thread_sleep(double sec);
  * buffer of the given size. That is, its length should be <= buffer size - 1,
  * because one extra char is taken by the trailing '\0'.
  */
-API_EXPORT char *
+char *
 getenv_safe(const char *name, size_t buf_size, bool *exceeded);
 
 #if !defined(__cplusplus) && !defined(static_assert)
```

Reply:
Better use Lua C IMO. AFAIR `tnt_internal_symbol` was added for a different purpose. Using it for Lua FFI looks cumbersome. If we do need a C function to be called via Lua FFI, IMO it'd be better to prefix its name with `tt_` or `tnt_`.

Reply:
>  If we do need a C function to be called via Lua FFI, IMO it'd be better to prefix its name with tt_ or tnt_.

Should it be a wrapper, like `tnt_getenv_safe`, which calls `getenv_safe` internally?

Reply:
I'd use a Lua C function, as you did initially IIRC, but a `tnt_getenv_safe` would look good to me as well if you prefer to use FFI. Actually, I'm a bit lost regarding when we should use the `tnt_` prefix and when not - there's a plenty of functions in `extra/exports` that lack the prefix. Maybe, just add `getenv_safe` there without any wrappers? If we decide to use wrappers for FFI functions, we'd better do it for all exported functions at once.

Reply:
Let's stick with [this](https://github.com/tarantool/tarantool/pull/7807#discussion_r999194832) version then.
FFI function in `src/lua/env.lua`, no prefix.

Reply:
I don't mind. I agree that we have no consistency now and my ideas about a (possibly) better approach are just ideas at the moment.